visualization/app/codeCharta/state/store/fileSettings/blacklist/blacklist.service.ts

```diff
@@ -15,14 +19,30 @@ export interface BlacklistSubscriber {
 export class BlacklistService implements StoreSubscriber, FilesSelectionSubscriber {
 	private static BLACKLIST_CHANGED_EVENT = "blacklist-changed"
 
-	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
+	constructor(
+		private $rootScope: IRootScopeService,
+		private storeService: StoreService,
+		private codeMapPreRenderService: CodeMapPreRenderService,
+		private dialogService: DialogService
+	) {
 		StoreService.subscribe(this.$rootScope, this)
 		FilesService.subscribe(this.$rootScope, this)
 	}
 
 	public onStoreChanged(actionType: string) {
 		if (isActionOfType(actionType, BlacklistActions)) {
-			this.notify(this.select())
+			if (
+				!NodeDecorator.doesExclusionResultInEmptyMap(
+					this.codeMapPreRenderService.getRenderMap(),
+					this.storeService.getState().fileSettings.blacklist
+				)
+			) {
+				this.storeService.dispatch(setIsLoadingMap(true))
+				this.notify(this.select())
+			} else {
+				this.storeService.dispatch(removeLastBlacklistItem(), true)
+				this.dialogService.showErrorDialog("Excluding all buildings is not possible.", "Blacklist Error")
+			}
```

Comment:
Is it possible to combine the logic in a way that it actually does remove all elements first, then you check if there are still elements and if there are, it's done, but if there are not, then you undo the last operation?

That would be more efficient :)

Reply:
If you do it that way, you'll first trigger that the blacklist has changed. Based on that event a lot of different services and components change their state, e.g. we calculate new metric data, since the maxValue might have changed. This would also trigger the re-decoration of the map.

We'd need to clone the previous map first in order to reset it, if all buildings will be excluded after decoration. We'd also need to notify all components and service, that they need to update their state again to the previous state.

Reply:
I personally would still do this, if it's reversible. It would be an optimization for the common case and a slow case for the case all folders are removed. This comment should not be considered blocking though.

Reply:
But this would require a clone. Cloning the map has the same complexity as checking if each building is excluded.

Reply:
Just for me to understand: how come we'd have to clone the map in that case? I would expect to actually change the map and then to reverse that action if we realize it runs into an issue (no files left).

Reply:
Well, cloning the map would be the easiest implementation. But now that I thought about this again:
While we update the map, we could add entries to a HashMap. Each entry represents the path of a node and it's previous state before we modified it.
If we run into a problem, we just traverse the tree again and check, if the path is in the HashMap and restore the previous state. We also remove the entry from the HashMap. Once the HashMap is empty, we're done.

I'll give this a shot

Reply:
I tried my best and pushed a solution, but I encounterd some issues that are quite hard to explain without code as a context.

Listening to blacklist-changes and as a reaction trigger a blacklist-change was tricky, since we block all other subscribers. e.g:

Blacklist all nodes -> store changes -> blacklist-changes -> trigger decoration -> dispatch blacklist-action -> store changes -> blacklist-changes -> Now the second blacklist-change is applied to all subscribers first and then we can execute the blacklist-change we triggered the first time.

This resulted in wrong data being displayed and wrong metrics being selected. I fixed this, by listening to the store directly and check for blacklist-events:

Blacklist all nodes -> store changes -> check if it was because of blacklist -> trigger decoration -> dispatch blacklist-action -> store changes -> blacklist-changes -> blacklist-changes from the first store-change.